binary_transparency/firmware/cmd/flash_tool/impl/flash_tool.go

```diff
@@ -224,13 +236,65 @@ func verifyAnnotations(c *client.ReadonlyClient, pb api.ProofBundle, fwMeta api.
 		return fmt.Errorf("failed to verify update with map checkpoint: %w", err)
 	}
 
-	// TODO(mhutchinson): Check the inclusion proof and use the values returned by the map.
-	afw, _, err := mc.Aggregation(mcp, pb.InclusionProof.LeafIndex)
+	// Get the aggregation and proof, and then check everything about it.
+	// This is a little paranoid as the inclusion proof is generated client-side.
+	// The pretense here is that there is a trust boundary between the code in this class,
+	// and everything else. In a production system, it is likely that the proof generation
+	// would live elsewhere (e.g. in the OTA packaging process), and the shorter proof
+	// bundle would be provided to the device.
+	afw, ip, err := mc.Aggregation(ctx, mcp.Revision, pb.InclusionProof.LeafIndex)
 	if err != nil {
 		return fmt.Errorf("failed to get map value for %q: %w", pb.InclusionProof.LeafIndex, err)
 	}
+	// 1. Check: the proof is for the correct key.
+	kbs := sha512.Sum512_256([]byte(fmt.Sprintf("summary:%d", pb.InclusionProof.LeafIndex)))
+	if !bytes.Equal(ip.Key, kbs[:]) {
+		return fmt.Errorf("received inclusion proof for key %x but wanted %x", ip.Key, kbs[:])
+	}
+	// 2. Check: the proof is for the correct value.
+	preimage, err := json.Marshal(afw)
```

Comment:
This might be brittle, maybe `mc.Aggregation` should return the raw leaf bytes and leave the parsing to the caller?

Reply:
There are comments in to this effect, but I will just resolve them now I guess. I went with this slightly risque approach because it led to a neat implementation of `fetch` inside the client that avoided needing to mess about with io closers when reading the body.

Reply:
This is addressed

Reply:
Yeah, I spotted them a bit further down, but I figured this was at least worth a discussion, and maybe I'd try to convince you to change it :)